src/EventSubscriber/APIToolkitService.php

```diff
@@ -2,143 +2,136 @@
+    private array $credentials = [];
+    private \SplObjectStorage $startTimes;
+
+    public function __construct(
+        private string $apiKey,
+        private string $rootURL = 'https://app.apitoolkit.io',
+    ) {
+        $this->startTimes = new \SplObjectStorage();
     }
 
-    $payload = (object)[
-      "duration" => round(hrtime(true) - $startTime),
-      "host" => $request->getHttpHost(),
-      "method" => $request->getMethod(),
-      "project_id" => $projectId,
-      "proto_major" => 1,
-      "proto_minor" => 1,
-      "query_params" => $query_params,
-      "path_params" => $request->attributes->get('_route_params'),
-      "raw_url" => $request->getRequestUri(),
-      "referer" => $request->headers->get("referer") ?? "",
-      "request_body" => base64_encode($request->getContent()),
-      "request_headers" => $request->headers->all(),
-      "response_body" => base64_encode($response->getContent()),
-      "response_headers" => $response->headers->all(),
-      "sdk_type" => "PhpSymfony",
-      "status_code" => $response->getStatusCode(),
-      "timestamp" => (new DateTime())->format("c"),
-      "url_path" => $request->attributes->get('_route'),
-    ];
-    return $payload;
-  }
-
-  // publishMessage leverages the credentials object to build a google pubsub topic 
-  // and then use that pubsub topic to publish a message into the APIToolkit pubsub queue.
-  public function publishMessage($payload, $credentials)
-  {
-    $pubsubClient = new PubSubClient([
-      "keyFile" => $credentials["pubsubKeyFile"]
-    ]);
-    $topic = $pubsubClient->topic($credentials["topic"]);
-    if (!$topic) return;
-
-    $topic->publish([
-      "data" => json_encode($payload, JSON_UNESCAPED_SLASHES)
-    ]);
-  }
-
-  public static function credentials($url, $api_key)
-  {
-    $url = $url . "/api/client_metadata";
-
-    $curlInit = curl_init($url);
-    curl_setopt($curlInit, CURLOPT_URL, $url);
-    curl_setopt($curlInit, CURLOPT_RETURNTRANSFER, true);
-
-    $headers = array(
-      "Authorization: Bearer $api_key",
-    );
-
-    curl_setopt($curlInit, CURLOPT_HTTPHEADER, $headers);
-    curl_setopt($curlInit, CURLOPT_SSL_VERIFYPEER, false);
-
-    $curlResponse = curl_exec($curlInit);
-
-    $response = json_decode($curlResponse, 1);
-    if ($curlResponse == false) {
-      curl_error($curlInit);
+    public static function getSubscribedEvents(): array
+    {
+        return [
+            KernelEvents::REQUEST => [
+                ['onKernelRequestPre', 10],
+            ],
+            KernelEvents::RESPONSE => [
+                ['onKernelResponsePre', 10],
+            ],
+        ];
     }
 
-    curl_close($curlInit);
+    public function onKernelRequestPre(RequestEvent $event): void
+    {
+        $this->startTimes[$event->getRequest()] = hrtime(true);
+    }
+
+    public function onKernelResponsePre(ResponseEvent $event): void
+    {
+        $credentials = $this->getCredentials();
```

Comment:
Thanks for the changes, @Metabor. But it appears you took out the caching. 
Won't this mean that we're making a http request for credentials on every single request on the server? 

Because PHP doesn't persist it between requests. This would tremendously slow down the traffic of users, since all their requests would be much much slower as a result.